modules/core/src/main/java/org/apache/ignite/internal/client/thin/TcpClientChannel.java

```diff
@@ -317,31 +356,51 @@ private void processNextResponse() throws ClientProtocolError, ClientConnectionE
                     lsnr.accept(this);
             }
 
+            if ((flags & ClientFlag.NOTIFICATION) != 0) {
+                short notificationCode = dataInput.readShort();
+
+                notificationOp = ClientOperation.fromCode(notificationCode);
+
+                if (notificationOp == null || !notificationOp.isNotification())
+                    throw new ClientProtocolError(String.format("Unexpected notification code [%d]", notificationCode));
+            }
+
             if ((flags & ClientFlag.ERROR) != 0)
                 status = dataInput.readInt();
         }
         else
             status = dataInput.readInt();
 
-        int hdrSize = (int)(dataInput.totalBytesRead() - bytesReadOnStartReq);
+        int hdrSize = (int)(dataInput.totalBytesRead() - bytesReadOnStartMsg);
+
+        byte[] res = null;
+        Exception err = null;
 
         if (status == 0) {
-            if (resSize <= hdrSize)
-                pendingReq.onDone();
-            else
-                pendingReq.onDone(dataInput.read(resSize - hdrSize));
+            if (msgSize > hdrSize)
+                res = dataInput.read(msgSize - hdrSize);
         }
+        else if (status == ClientStatus.SECURITY_VIOLATION)
+            err = new ClientAuthorizationException();
         else {
-            resIn = new BinaryHeapInputStream(dataInput.read(resSize - hdrSize));
+            resIn = new BinaryHeapInputStream(dataInput.read(msgSize - hdrSize));
 
-            String err = new BinaryReaderExImpl(null, resIn, null, true).readString();
+            String errMsg = new BinaryReaderExImpl(null, resIn, null, true).readString();
 
-            switch (status) {
-                case ClientStatus.SECURITY_VIOLATION:
-                    pendingReq.onDone(new ClientAuthorizationException());
-                default:
-                    pendingReq.onDone(new ClientServerError(err, status, resId));
-            }
+            err = new ClientServerError(errMsg, status, resId);
+        }
+
+        if (notificationOp == null) { // Respone received.
+            ClientRequestFuture pendingReq = pendingReqs.get(resId);
+
+            if (pendingReq == null)
+                throw new ClientProtocolError(String.format("Unexpected response ID [%s]", resId));
+
+            pendingReq.onDone(res, err);
+        }
+        else { // Notification received.
+            for (NotificationListener lsnr : notificationLsnrs)
+                lsnr.acceptNotification(this, notificationOp, resId, res, err);
```

Comment:
Minor comment: TcpClientChannel always have only one listener, right? Why do we need collection here?

Reply:
Yes, currently we need only one listener, but IMO with the collection of listeners API looks more consistent and it costs almost nothing comparing to one listener.